File: src/main/java/xyz/chunkstories/core/generator/FlatGenerator.java
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.generator;

import xyz.chunkstories.api.voxel.Voxel;
import xyz.chunkstories.api.world.World;
import xyz.chunkstories.api.world.chunk.Chunk;
import xyz.chunkstories.api.world.generator.WorldGenerator;
import xyz.chunkstories.api.world.generator.WorldGeneratorDefinition;

import java.util.Random;

public class FlatGenerator extends WorldGenerator {
	Random rnd = new Random();

	public FlatGenerator(WorldGeneratorDefinition type, World world) {
		super(type, world);
		worldsize = world.getSizeInChunks() * 32;

		this.GROUND_VOXEL = world.getGameContext().getContent().getVoxels().getVoxel("grass");
		this.WALL_VOXEL = world.getGameContext().getContent().getVoxels().getVoxel("cobble");
		this.WALL_TOP_VOXEL = world.getGameContext().getContent().getVoxels().getVoxel("ironGrill");
	}

	int worldsize;
	private final Voxel GROUND_VOXEL;
	private final Voxel WALL_VOXEL;
	private final Voxel WALL_TOP_VOXEL;

	@Override
	public void generateChunk(Chunk chunk) {
		int cx = chunk.getChunkX();
		int cy = chunk.getChunkY();
		int cz = chunk.getChunkZ();

		rnd.setSeed(cx * 32 + cz + 48716148);

		for (int x = 0; x < 32; x++)
			for (int z = 0; z < 32; z++) {
				Voxel type = WALL_VOXEL; // cobble

				int v = 21; // base height
				if ((cx * 32 + x) % 256 == 0 || (cz * 32 + z) % 256 == 0) {
					v = 30; // wall height
				} else
					type = GROUND_VOXEL;
				// int v = 250;
				int y = cy * 32;
				while (y < cy * 32 + 32 && y <= v) {
					if (y == 30)
						type = WALL_TOP_VOXEL;
					chunk.pokeSimpleSilently(x, y, z, type, -1, -1, 0);
					y++;
				}
			}
	}
}

File: src/main/java/xyz/chunkstories/core/generator/FlatGenerator.kt
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.generator

import xyz.chunkstories.api.block.BlockType
import xyz.chunkstories.api.content.json.asInt
import xyz.chunkstories.api.world.World
import xyz.chunkstories.api.world.cell.CellData
import xyz.chunkstories.api.world.chunk.Chunk
import xyz.chunkstories.api.world.generator.WorldGenerator
import xyz.chunkstories.api.world.generator.WorldGeneratorDefinition

import java.util.Random

class FlatGenerator(definition: WorldGeneratorDefinition, world: World) : WorldGenerator(definition, world) {
    internal var rnd = Random()

    private var worldsize: Int = 0

    private val cellSize: Int

    private val GROUND_VOXEL: BlockType
    private val WALL_VOXEL: BlockType
    private val WALL_TOP_VOXEL: BlockType

    init {
        worldsize = world.properties.size.sizeInChunks * 32

        this.GROUND_VOXEL = world.gameInstance.content.blockTypes.get("grass")!!
        this.WALL_VOXEL = world.gameInstance.content.blockTypes.get("cobble")!!
        this.WALL_TOP_VOXEL = world.gameInstance.content.blockTypes.get("iron_bars")!!

        this.cellSize = definition.properties["cellSize"].asInt ?: 0
    }

    override fun generateWorldSlice(chunks: Array<PreChunk>) {
        for (chunkY in chunks.indices) {
            generateChunk(chunks[chunkY])
        }
    }

    private fun generateChunk(chunk: PreChunk) {
        val cx = chunk.chunkX
        val cy = chunk.chunkY
        val cz = chunk.chunkZ

        rnd.setSeed((cx * 32 + cz + 48716148).toLong())

        for (x in 0..31)
            for (z in 0..31) {
                var type = WALL_VOXEL // cobble

                var terrainHeight = 21

                if (cellSize != 0 && onWall(cx * 32 + x, cz * 32 + z)) {
                    terrainHeight = 30
                } else {
                    type = GROUND_VOXEL
                }

                var y = cy * 32
                while (y < cy * 32 + 32 && y <= terrainHeight) {
                    if (y == 30)
                        type = WALL_TOP_VOXEL
                    chunk.setCellData(x, y, z, CellData(type))
                    y++
                }
            }
    }

    override fun generateWorldSlicePhaseII(chunks: Array<Chunk>) {

    }

    private fun onWall(x: Int, z: Int) = ((x) % 256 == 0 || (z) % 256 == 0)
}
